engine/crates/integration-tests/tests/federation/basic/headers.rs

```diff
@@ -141,38 +139,112 @@ fn test_subgraph_specific_header_forwarding() {
             .await
     });
 
-    let mut response: Response = serde_json::from_value(response.into_data()).unwrap();
-    response.headers.sort();
-
-    insta::assert_debug_snapshot!(response.headers, @r###"
-    [
-        Header {
-            name: "accept",
-            value: "*/*",
-        },
-        Header {
-            name: "content-type",
-            value: "application/json",
-        },
-        Header {
-            name: "x-foo",
-            value: "BAR",
-        },
-        Header {
-            name: "x-forwarded",
-            value: "boom",
-        },
-    ]
+    insta::assert_json_snapshot!(response, @r###"
+    {
+      "data": {
+        "headers": [
+          {
+            "name": "accept",
+            "value": "application/json"
+          },
+          {
+            "name": "content-length",
+            "value": "78"
+          },
+          {
+            "name": "content-type",
+            "value": "application/json"
+          },
+          {
+            "name": "x-foo",
+            "value": "BAR"
+          },
+          {
+            "name": "x-forwarded",
+            "value": "boom"
+          }
+        ]
+      }
+    }
     "###);
 }
 
+#[test]
+fn should_not_propagate_blacklisted_headers() {
+    runtime().block_on(async move {
+        let echo_mock = MockGraphQlServer::new(EchoSchema).await;
+
+        let engine = Engine::builder()
+            .with_subgraph("echo", &echo_mock)
+            .with_header_rule(SubgraphHeaderRule::Forward(SubgraphHeaderForward {
+                name: NameOrPattern::Pattern(Regex::new(".*").unwrap()),
+                default: None,
+                rename: None,
+            }))
+            .with_header_rule(SubgraphHeaderRule::Insert(SubgraphHeaderInsert {
+                name: "Content-Type".into(),
+                value: "application/trust-me".into(),
+            }))
+            .with_header_rule(SubgraphHeaderRule::RenameDuplicate(SubgraphRenameDuplicate {
+                name: "User-Agent".into(),
+                default: None,
+                rename: "TE".into(),
+            }))
+            .build()
+            .await;
+
+        let response = engine
+            .execute("query { headers { name value }}")
+            .header("User-Agent", "Rusty")
+            .header("Accept", "application/json")
+            .header("Accept-Encoding", "gzip")
+            .header("Accept-Charset", "utf-8")
+            .header("Accept-Ranges", "bytes")
+            .header("Content-Length", "728")
+            .header("Content-Type", "application/jpeg")
```

Comment:
Kind of seems like a bug that we respond to requests with `Content-Type: application/jpeg` tbh.  Possibly the same for some of these other ones.

Reply:
(Not saying we should fix it right now, just an observation)

Reply:
what do you mean? the response shows the headers the subgraph received, so `application/json`

Reply:
ahhh ok got what you mean, `Content-Type` for the request is handled by `axum` since it deserializes the JSON.

Reply:
so yeah wouldn't be able to test that we don't forward this header with the gateway

Reply:
Oh right of course - I forgot we're skipping the whole HTTP layer in these tests.  Cool, never mind then.